SpatialDeconvolution.cu

```diff
@@ -0,0 +1,369 @@
+    // Force batch
+    batch = 0;
+    THCudaTensor_resize4d(state, input, 1, input->size[0], input->size[1], input->size[2]);
+  } else {
+    luaL_argcheck(L, input->size[1] == nInputPlane, 2, "input channels and nInputPlane dont match");
+  }
+
+  long inputWidth   = input->size[3];
+  long inputHeight  = input->size[2];
+  long outputWidth  = (inputWidth - 1) * dW - 2*padW + kW;
+  long outputHeight = (inputHeight - 1) * dH - 2*padH + kH;
+
+  // Batch size + input planes
+  long batchSize = input->size[0];
+
+  // Resize output
+  THCudaTensor_resize4d(state, output, batchSize, nOutputPlane, outputHeight, outputWidth);
+
+  // Resize temporary columns
+  THCudaTensor_resize2d(state, columns, nOutputPlane*kW*kH, inputHeight*inputWidth);
+
+  // Define a buffer of ones, for bias accumulation
+  // Note: this buffer can be shared with other modules, it only ever gets increased,
+  // and always contains ones.
+  if (ones->nDimension != 2 || ones->size[0]*ones->size[1] < outputHeight*outputWidth) {
+    // Resize plane and fill with ones...
+    THCudaTensor_resize2d(state, ones, outputHeight, outputWidth);
+    THCudaTensor_fill(state, ones, 1);
+  }
+
+  // Helpers
+  THCudaTensor *input_n = THCudaTensor_new(state);
+  THCudaTensor *output_n = THCudaTensor_new(state);
+
+  // For each elt in batch, do:
+  for (int elt = 0; elt < batchSize; elt ++) {
+    // Matrix mulitply per output:
+    THCudaTensor_select(state, input_n, input, 0, elt);
+    THCudaTensor_select(state, output_n, output, 0, elt);
+
+    // M,N,K are dims of matrix A and B
+    // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+    long m = weight->size[1];
+    long n = columns->size[1];
+    long k = weight->size[0];
+
+    // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+    THCudaBlas_gemm(
+        state,
+        'n', 't',
+        n, m, k,
+        1,
+        THCudaTensor_data(state, input_n), n,
+        THCudaTensor_data(state, weight), m,
+        0,
+        THCudaTensor_data(state, columns), n
+    );
+
+    // Unpack columns back into input:
+    col2im(
+      THCState_getCurrentStream(state),
+      THCudaTensor_data(state, columns),
+      nOutputPlane, outputHeight, outputWidth, kH, kW, padH, padW, dH, dW,
+      THCudaTensor_data(state, output_n)
+    );
+
+    // Do Bias after:
+    // M,N,K are dims of matrix A and B
+    // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+    long m_ = nOutputPlane;
+    long n_ = outputHeight * outputWidth;
+    long k_ = 1;
+
+    // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+    THCudaBlas_gemm(
+        state,
+        't', 'n',
+        n_, m_, k_,
+        1,
+        THCudaTensor_data(state, ones), k_,
+        THCudaTensor_data(state, bias), k_,
+        1,
+        THCudaTensor_data(state, output_n), n_
+    );
```

Comment:
I have a general question about 'Do Bias after':

Using the following notation (input+output defined as in a classic O = nn.SpatialConvolution():forward(I)):
ic, ih, iw = input channels, height, width
oc, oh, ow = output channels, height, width
kh, kw = filter kernel height, width

I = input of size (ic, ih * iw)
O = output of size (oc, oh * ow)
im2col(I) producing (ic * kh * kw, oh * ow)
W=weights of size (oc, ic * kh * kw)
b= bias of size (oc, 1)

A classic forward convolution might the look like:
O=W * im2col(I) + b * ones(1, oh * ow)

A deconvolution operation could be:
I=col2im(w^T * (O - b * ones(1, oh * ow))

What confuses me is the addition of the bias to the output:
If the bias is generated by the ctor of nn.SpatialConvolution (base class) we have only a bias vector with oc elements (=which corresponds to the number of input(!)-planes for deconvolution). During deconvolution the convolution-output gets the size of the 'original' input (= we produce ic planes).

Due to the swapping of nOutputPlane, nInputPlane in ln. 15 in this code the nOutputPlanes would be ic in my notation and nInputPlanes oc.

The matrix multiplication in ln. 107 reads from the bias k_ x m_ elements (1, ci) .. but if the bias was defined for for a forward convolution this vector should have oc elements - one entry per output channel (feature-map) of the forward conv.

@szagoruyko mentioned that he checked this impl against caffe and it seems they are also adding the bias afterwards: https://github.com/BVLC/caffe/blob/master/src/caffe/layers/deconv_layer.cpp#L36 (@fmassa thnx for the link).

Swapping nInputPlanes and nOutputPlanes + api expecting column-major matrices + deconv becomes quickly more challenging than I had expected.  Probably I just got something wrong. It took me some time to untangle the bits of Deconv. It would be great if somebody could help me a little bit by checking/correcting my 'math'. :-)  thx